## 데이터 분석과 머신러닝 실습 3 : 분류(Classification)

<div style="text-align: right"> 작성자 : 데이터사이언스연구소 유광명, 이영생, 임혜원 </div>  

---

### 실습 3.1 IRIS(붓꽃) 품종 분류 문제
---
붓꽃의 품종을 분류하는 머신러닝 모델 만들기

<img src="img/iris.png" width="500">

**1) 첫번째 머신러닝 모델**  
로지스틱 회귀(Logistic regression) 분류 모델

**라이브러리 import**

In [1]:
import numpy as np #numpy
import pandas as pd #pandas

**데이터 읽어오기 및 기본정보 확인**  
pandas의 read_csv 메서드를 사용하여 데이터 셋을 iris 변수에 저장한다.

In [2]:
iris = pd.read_csv('data/iris.data', header = None)

In [3]:
type(iris)

pandas.core.frame.DataFrame

데이터는 DataFrame 형식이다.

head() 메서드로 데이터의 최초 5개 샘플을 확인한다.

In [4]:
iris.head()

0    1    2    3            4
0  5.1  3.5  1.4  0.2  Iris-setosa
1  4.9  3.0  1.4  0.2  Iris-setosa
2  4.7  3.2  1.3  0.2  Iris-setosa
3  4.6  3.1  1.5  0.2  Iris-setosa
4  5.0  3.6  1.4  0.2  Iris-setosa

In [5]:
iris.head(10)

0    1    2    3            4
0  5.1  3.5  1.4  0.2  Iris-setosa
1  4.9  3.0  1.4  0.2  Iris-setosa
2  4.7  3.2  1.3  0.2  Iris-setosa
3  4.6  3.1  1.5  0.2  Iris-setosa
4  5.0  3.6  1.4  0.2  Iris-setosa
5  5.4  3.9  1.7  0.4  Iris-setosa
6  4.6  3.4  1.4  0.3  Iris-setosa
7  5.0  3.4  1.5  0.2  Iris-setosa
8  4.4  2.9  1.4  0.2  Iris-setosa
9  4.9  3.1  1.5  0.1  Iris-setosa

타겟 변수만 살펴보기(5번째 컬럼)

In [6]:
iris[4].head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: 4, dtype: object

In [7]:
type(iris[4])

pandas.core.series.Series

단일 컬럼만 인덱싱 하면 series 타입으로 리턴된다.

unique 메서드로 타겟(또는 클래스)의 종류 확인한다.

In [8]:
iris[4].unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

타겟의 종류는 3가지 이다.

In [9]:
iris.shape

(150, 5)

150x5, 2차원 데이터 셋이다.

**데이터 전처리**  
데이터 셋을 입력(피처)과 출력(타겟)으로 나눈다.

In [10]:
X = iris.iloc[:, 0:3]
y = iris.iloc[:, 4]

데이터 셋이 잘 분리되었는지 확인한다.

In [11]:
X.head()

0    1    2
0  5.1  3.5  1.4
1  4.9  3.0  1.4
2  4.7  3.2  1.3
3  4.6  3.1  1.5
4  5.0  3.6  1.4

In [12]:
y.head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: 4, dtype: object

**모델학습(Training)**  
로지스틱 회귀(Logistic Regression) 모델을 생성하고, 데이터 셋으로 학습시킨다.

<img src="img/logistic_regression.png" width="500">

In [13]:
from sklearn.linear_model import LogisticRegression

In [14]:
logreg = LogisticRegression() # 모델생성

In [15]:
logreg.fit(X, y) # 모델학습

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

**성능평가(Evaluation)**

피처 데이터만 입력하여 예측결과를 저장한다.

In [16]:
y_pred = logreg.predict(X)

In [17]:
y_pred[:5]

array(['Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa'], dtype=object)

실제 결과와 예측 결과를 비교하여 평균을 계산한다.

In [18]:
np.mean(y_pred==y)

0.9333333333333333

모델의 score 메서드를 사용해서 동일한 결과를 얻을수도 있다.

In [19]:
logreg.score(X, y)

0.9333333333333333

**2) 데이터셋을 분리하여 학습/ 검증하기**  
로지스틱 회귀(Logistic regression)

<img src="img/train_test.png" width="400">

**데이터 읽기**

In [20]:
iris = pd.read_csv('data/iris.data', header = None)

**데이터 전처리 및 학습/ 시험 데이터 나누기**  
- 데이터 셋을 입력(피처)과 출력(타겟)으로 나눈다.
- 학습 데이터와 시험(검증) 데이터로 분리한다.

In [21]:
X = iris.iloc[:, 0:3]
y = iris.iloc[:, 4]

사이킷런의 train_test_split 라이브러리를 사용하여 학습, 시험 데이터로 분리한다.

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                    stratify = y, random_state = 7,test_size = 0.3)

In [24]:
print('X_train :',X_train.shape)
print('y_train :',y_train.shape)
print('X_test :',X_test.shape)
print('y_test :',y_test.shape)

X_train : (105, 3)
y_train : (105,)
X_test : (45, 3)
y_test : (45,)


**모델학습(Training)**  
학습용 데이터(X_train, y_train)로 모델을 학습시킨다.

In [25]:
from sklearn.linear_model import LogisticRegression

In [26]:
logreg = LogisticRegression(random_state=7)

In [27]:
logreg.fit(X_train, y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=7, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

**성능평가(Evaluation)**  
시험용 데이터(X_test, y_test)로 모델 성능을 평가한다.

In [28]:
logreg.score(X_train, y_train)

0.9333333333333333

In [29]:
logreg.score(X_test, y_test)

0.9111111111111111

- 학습용 데이터 성능평가 결과 :93.33%  
- 시험용 데이터 성능평가 결과 :91.11%

predict_proba 메서드를 사용하면 각 클래스의 예측확률을 확인할 수 있다.

In [30]:
proba = logreg.predict_proba(X_test)
proba

array([[2.97544814e-01, 6.49106544e-01, 5.33486425e-02],
       [8.37628637e-02, 7.56095593e-01, 1.60141543e-01],
       [3.73246557e-04, 3.91214781e-01, 6.08411972e-01],
       [8.95529739e-01, 1.04372594e-01, 9.76676062e-05],
       [1.34712954e-01, 6.76583218e-01, 1.88703829e-01],
       [9.14980136e-01, 8.49273470e-02, 9.25169319e-05],
       [8.24429524e-03, 4.67460769e-01, 5.24294936e-01],
       [2.13665749e-03, 3.76641680e-01, 6.21221663e-01],
       [5.28255236e-03, 4.66556397e-01, 5.28161051e-01],
       [9.14840677e-01, 8.51501788e-02, 9.14406556e-06],
       [8.43453578e-01, 1.56274100e-01, 2.72321581e-04],
       [4.44365314e-03, 3.16329863e-01, 6.79226484e-01],
       [9.12266808e-01, 8.74152589e-02, 3.17933047e-04],
       [2.22038592e-02, 5.03460180e-01, 4.74335961e-01],
       [6.80656854e-02, 4.64633383e-01, 4.67300931e-01],
       [4.42345993e-02, 7.48431933e-01, 2.07333468e-01],
       [1.05346976e-03, 3.09329206e-01, 6.89617324e-01],
       [7.83473293e-02, 7.21118

### 실습 3.2 와인 등급 분류 문제
---
와인의 등급을 분류하는 머신러닝 모델 만들기

<img src="img/wine.png" width="400">

**라이브러리 import**

In [31]:
import numpy as np #numpy
import pandas as pd #pandas

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

**데이터 읽어오기 및 기본정보 확인**  
pandas의 read_csv 메서드를 사용하여 데이터 셋을 wine 변수에 저장

In [32]:
wine = pd.read_csv('data/wine.data')

In [33]:
wine.head()

Class label  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0            1    14.23        1.71  2.43               15.6        127   
1            1    13.20        1.78  2.14               11.2        100   
2            1    13.16        2.36  2.67               18.6        101   
3            1    14.37        1.95  2.50               16.8        113   
4            1    13.24        2.59  2.87               21.0        118   

   Total phenols  Flavanoids  Nonflavnoid phenols  Proanthocyanins  \
0           2.80        3.06                 0.28             2.29   
1           2.65        2.76                 0.26             1.28   
2           2.80        3.24                 0.30             2.81   
3           3.85        3.49                 0.24             2.18   
4           2.80        2.69                 0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

타겟 컬럼은 'Class label'이다.

In [34]:
wine['Class label'].unique()

array([1, 2, 3], dtype=int64)

타겟의 종류는 3가지 이다. 

In [35]:
wine.shape

(178, 14)

178x14, 2차원 데이터 셋이다

**데이터 전처리 및 학습/ 시험 데이터 나누기**  
- 데이터 셋을 입력(피처)과 출력(타겟)으로 나눈다.
- 학습 데이터와 시험(검증) 데이터로 분리한다.

In [36]:
X = wine.iloc[:, 1:14]
y = wine.iloc[:, 0]

In [37]:
X.head()

Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  Total phenols  \
0    14.23        1.71  2.43               15.6        127           2.80   
1    13.20        1.78  2.14               11.2        100           2.65   
2    13.16        2.36  2.67               18.6        101           2.80   
3    14.37        1.95  2.50               16.8        113           3.85   
4    13.24        2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavnoid phenols  Proanthocyanins  Color intensity   Hue  \
0        3.06                 0.28             2.29             5.64  1.04   
1        2.76                 0.26             1.28             4.38  1.05   
2        3.24                 0.30             2.81             5.68  1.03   
3        3.49                 0.24             2.18             7.80  0.86   
4        2.69                 0.39             1.82             4.32  1.04   

   OD280/OD315 of diluted wines  Proline  
0                          3.92     1065  
1                          3.40     1050  
2                          3.17     1185  
3                          3.45     1480  
4                          2.93      735

In [38]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: Class label, dtype: int64

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                    stratify = y, random_state = 7,test_size = 0.3)

**모델학습(Training)**  
학습용 데이터(X_train, y_train)로 모델을 학습시킨다.

In [40]:
logreg = LogisticRegression(random_state=7)

In [41]:
logreg.fit(X_train, y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=7, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

**성능평가(Evaluation)**  
시험용 데이터(X_test, y_test)로 모델 성능을 평가한다.

In [42]:
logreg.score(X_train, y_train)

0.9919354838709677

In [43]:
logreg.score(X_test, y_test)

0.9444444444444444